File: android/app/src/main/java/com/prince/turtlekeyboard/integration/poll/PollIntegration.java
package com.prince.turtlekeyboard.integration.poll;

import android.util.Log;
import android.view.inputmethod.EditorInfo;

import androidx.annotation.Nullable;

import com.prince.kbd.core.AssetPrompts;
import com.prince.kbd.core.CommandSpec;
import com.prince.kbd.core.GeminiService;
import com.prince.kbd.core.IntegrationContext;
import com.prince.kbd.core.IntegrationSession;
import com.prince.kbd.core.KeyboardIntegration;
import com.prince.kbd.core.SheetViewFactory;
import com.prince.turtlekeyboard.ai.AiErrorMessages;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * {@code /poll} — shape a terse user prompt into a {@code {question, options}} object
 * via Gemini, write the poll doc via {@link RealtimePollClient#createPoll}, then commit
 * the shareable App Link URL into the host editor.
 */
public class PollIntegration implements KeyboardIntegration {

    private static final String TAG = "PollIntegration";

    public static final String ROUTE_KEY = "poll";

    private static final long BUSY_BANNER_MS = 30_000L;
    private static final long FAIL_BANNER_MS = 2_500L;
    private static final long EMPTY_BANNER_MS = 2_200L;

    @Override public String id() { return "poll"; }

    @Override
    @Nullable
    public IntegrationSession activate(EditorInfo info, IntegrationContext ctx) {
        return null;
    }

    @Override
    public List<CommandSpec> commands() {
        return Collections.singletonList(
                new CommandSpec("poll", "Poll", "📊", true, this::handlePoll)
        );
    }

    @Override
    public Map<String, SheetViewFactory> sheetRoutes() {
        return Collections.singletonMap(ROUTE_KEY, PollSheetView::new);
    }

    // -- handler -------------------------------------------------------------

    private void handlePoll(String prompt, IntegrationContext ctx) {
        final String trimmed = prompt == null ? "" : prompt.trim();
        if (trimmed.isEmpty()) {
            ctx.showBanner("What's the poll? e.g. /poll best dinner spot", EMPTY_BANNER_MS);
            return;
        }
        String systemPrompt = AssetPrompts.load(ctx.appContext(), "poll");
        if (systemPrompt.isEmpty()) {
            ctx.showBanner("Poll prompt missing — clean rebuild needed", FAIL_BANNER_MS);
            return;
        }
        ctx.showBanner("Creating poll…", BUSY_BANNER_MS);
        ctx.ai().text(systemPrompt, trimmed, new GeminiService.TextCallback() {
            @Override public void onText(String text) { onModelText(ctx, text); }
            @Override public void onError(String reason) {
                Log.w(TAG, "poll ai.text failed: " + reason);
                ctx.showBanner(AiErrorMessages.userMessage(reason), FAIL_BANNER_MS);
            }
        });
    }

    private void onModelText(IntegrationContext ctx, String rawJson) {
        String stripped = stripCodeFences(rawJson);
        JSONObject parsed;
        try {
            parsed = new JSONObject(stripped);
        } catch (Exception e) {
            Log.w(TAG, "poll JSON parse failed; raw=" + stripped, e);
            ctx.showBanner("Couldn't shape that into a poll — try a clearer prompt", FAIL_BANNER_MS);
            return;
        }
        final String question = parsed.optString("question", "").trim();
        JSONArray optsArr = parsed.optJSONArray("options");
        if (question.isEmpty() || optsArr == null || optsArr.length() < 2) {
            ctx.showBanner("Couldn't shape that into a poll — try a clearer prompt", FAIL_BANNER_MS);
            return;
        }
        final List<String> options = new ArrayList<>(optsArr.length());
        for (int i = 0; i < optsArr.length(); i++) {
            String label = optsArr.optString(i, "").trim();
            if (!label.isEmpty()) options.add(label);
        }
        if (options.size() < 2) {
            ctx.showBanner("Couldn't shape that into a poll — try a clearer prompt", FAIL_BANNER_MS);
            return;
        }
        RealtimePollClient.createPoll(question, options, new RealtimePollClient.CreateCallback() {
            @Override public void onSuccess(RealtimePollClient.CreateResult result) {
                ctx.commitText(result.url);
            }
            @Override public void onError(String reason) {
                Log.w(TAG, "RealtimePollClient.createPoll failed: " + reason);
                ctx.showBanner(bannerForError(reason), FAIL_BANNER_MS);
            }
        });
    }

    private static String bannerForError(String code) {
        switch (code) {
            case "not_signed_in":
                return "Open Turtle and sign in to create polls";
            case "invalid_payload":
                return "Couldn't shape that into a poll — try a clearer prompt";
            case "network":
                return "Poll create failed — check your connection";
            case "permission_denied":
                return "Poll create blocked — try signing out and back in";
            default:
                return "Poll create failed: " + code;
        }
    }

    /** Models occasionally wrap JSON in ```json … ``` despite the prompt. */
    private static String stripCodeFences(String s) {
        if (s == null) return "";
        String t = s.trim();
        if (t.startsWith("```")) {
            int firstNl = t.indexOf('\n');
            if (firstNl > 0) t = t.substring(firstNl + 1);
            int closing = t.lastIndexOf("```");
            if (closing >= 0) t = t.substring(0, closing);
        }
        return t.trim();
    }
}

File: ios/TurtleKeyboardExtension/Integration/Poll/PollIntegration.swift
import Foundation
#if os(iOS)
import UIKit

/// Owns the full `/poll` flow:
///   1. Route the shaping step through `CommandRouter`, which turns the
///      user's terse prompt into `{question, options}` JSON on the cheapest
///      tier that can do it — Apple's on-device model first (free, offline,
///      guided generation), cloud only if that's unavailable or the draft
///      fails validation.
///   2. POST to `turtle-worker` `/poll` via `PollClient`.
///   3. Commit the returned shareable App Link URL into the host field.
///
/// Mirrors Android's `PollIntegration`. The system prompt lives in
/// `commands/prompts/poll.txt` and is owned by `CommandRouter.systemPrompt`,
/// so both tiers shape polls the same way.
///
/// Note step 2 always hits the network: a poll is a shareable artifact, so
/// `InferenceMode.onDeviceOnly` keeps the *inference* local, not the poll.
final class PollIntegration: KeyboardIntegration {

    let id = "poll"

    /// URL route key — matches `https://www.turtlekeyboard.com/poll/<id>`.
    static let routeKey = "poll"

    private static let failBannerMs = 2_500
    private static let emptyBannerMs = 2_200

    func commands() -> [CommandSpec] {
        [
            CommandSpec(
                name: "poll", label: "Poll", emoji: "📊", needsPrompt: true,
                handler: { prompt, ctx in Self.handle(prompt: prompt, ctx: ctx) }
            ),
        ]
    }

    static func handle(prompt: String, ctx: IntegrationContext) {
        let trimmed = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            ctx.showBanner("What's the poll? e.g. /poll best dinner spot",
                           autoHideMs: emptyBannerMs)
            return
        }
        // The router falls back to an inline prompt if the asset is missing,
        // but a missing asset means the Run Script's inputPaths/outputPaths
        // aren't wired — say so rather than silently running the fallback.
        guard PromptLoader.load(id: "poll")?.isEmpty == false else {
            ctx.showBanner("Poll prompt missing — clean rebuild needed",
                           autoHideMs: failBannerMs)
            return
        }

        ctx.showBusy("Creating poll…")

        ctx.llm.complete(
            command: "poll",
            prompt: trimmed,
            onText: { text in
                handleModelText(rawJson: text, ctx: ctx)
            },
            onError: { reason in
                DispatchQueue.main.async {
                    ctx.showBanner("Poll failed: \(reason)", autoHideMs: failBannerMs)
                }
            }
        )
    }

    /// Parses the model's JSON output and POSTs to the Worker.
    private static func handleModelText(rawJson: String, ctx: IntegrationContext) {
        let stripped = stripCodeFences(rawJson)
        guard let data = stripped.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            DispatchQueue.main.async {
                ctx.showBanner("Couldn't shape that into a poll — try a clearer prompt",
                               autoHideMs: failBannerMs)
            }
            return
        }
        let question = (obj["question"] as? String ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let optsArr = (obj["options"] as? [Any]) ?? []
        let options: [String] = optsArr.compactMap { raw in
            let s = (raw as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            return s.isEmpty ? nil : s
        }
        guard !question.isEmpty, options.count >= 2 else {
            DispatchQueue.main.async {
                ctx.showBanner("Couldn't shape that into a poll — try a clearer prompt",
                               autoHideMs: failBannerMs)
            }
            return
        }

        Task {
            do {
                let result = try await PollClient.createPoll(question: question, options: options)
                await MainActor.run { ctx.commitText(result.url) }
            } catch {
                await MainActor.run {
                    ctx.showBanner("Couldn’t create the poll. Please try again.", autoHideMs: failBannerMs)
                }
            }
        }
    }

    /// Even when told not to, models occasionally wrap JSON in ```json … ```.
    private static func stripCodeFences(_ s: String) -> String {
        var t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.hasPrefix("```") {
            if let firstNl = t.firstIndex(of: "\n") {
                t = String(t[t.index(after: firstNl)...])
            }
            if let closing = t.range(of: "```", options: .backwards) {
                t = String(t[..<closing.lowerBound])
            }
        }
        return t.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
#endif
